lib/spack/spack/spec.py

```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+            # Only set it if it hasn't been spliced before
+            ancestor._build_spec = ancestor._build_spec or ancestor.copy()
+            ancestor.clear_cached_hashes(ignore=(ht.package_hash.attr,))
+
+        # For each direct dependent in the link/run graph, replace the dependency on
+        # node with one on replacement
+        # For each build dependent, restrict the edge to build-only
+        for edge in self.edges_from_dependents():
+            if edge.parent not in ancestors_in_context:
+                continue
+            build_dep = edge.depflag & dt.BUILD
+            other_dep = edge.depflag & ~dt.BUILD
+            if build_dep:
+                parent_edge = [e for e in edge.parent._dependencies[self.name] if e.spec is self]
+                assert len(parent_edge) == 1
+
+                edge.depflag = dt.BUILD
+                parent_edge[0].depflag = dt.BUILD
+            else:
+                edge.parent._dependencies.edges[self.name].remove(edge)
+                self._dependents.edges[edge.parent.name].remove(edge)
+
+            if other_dep:
+                edge.parent._add_dependency(replacement, depflag=other_dep, virtuals=edge.virtuals)
+
+    def get_analogues(self, candidates_generator, self_root, other_root):
+        """Find all specs in candidate_generator that are splice matches for self
+
+        self_root and other_root are passed as arguments to ``self._splie_match``."""
+        analogues = [
+            dep
+            for dep in candidates_generator
+            if self._splice_match(dep, self_root=self_root, other_root=other_root)
+        ]
+        if not analogues:
+            return []
+
+        name_analogues = [a for a in analogues if a.name == self.name]
+        return name_analogues or analogues
+
+    def _splice_helper(self, replacement, self_root, other_root):
+        """Main loop of a transitive splice.
+
+        Topo traversal of self ensures that if a node is unreachable in the end result, we will
+        never consider it.
+        For each node, find any analogue in replacement and swap it in.
+        We assume only build deps are handled outside of this method
+        """
+        ids = [id(s) for s in replacement.traverse()]
+
+        changed = True
+        while changed:
+            changed = False
+
+            # Intentionally allowing traversal to change on each iteration
+            # using breadth-first traversal to ensure we only reach nodes that will
+            # be in final result
+            for node in self.traverse(root=False, order="topo", deptype=dt.ALL & ~dt.BUILD):
+                # If this node has already been swapped in, don't consider it again
+                if id(node) in ids:
+                    continue
+                analogues = node.get_analogues(
+                    replacement.traverse(deptype=dt.ALL & ~dt.BUILD),
+                    self_root=self_root,
+                    other_root=other_root,
+                )
```

Comment:
(Looking further into how this can be improved is the only element of this review not addressed by what I'm pushing right now).

Reply:
You can still key by name bc I think you nearly always know that names of specs or providers are the first criteria for compatibility. The package/provider name identifies a list of specs — there can be multiple specs in there and you’d compare with all the specs in any given list instead of all the specs in the DAG. 

I *think* using a dict like this avoids O(n^2) match calls and focuses only on the ones that could possibly match, without losing generality.

We’ve talked about allowing arbitrary ABI compatibility rules between packages with h different names, but I don’t see right now how we’d do that without using virtuals or some other kind of ABI provider, so I *think* the approach won’t change much if we tweak how we do ABI matching.